### PR TITLE
Add flake8-boolean-trap

### PR DESCRIPTION
**Pull request by pwoolvett**
Disclaimer: I have no rust experience whatsoever.

### CODE REVIEW

File: src/flake8_boolean_trap/plugins.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                }
+                ExprKind::Constant { value, .. } => {
+                    hint = value.to_string();
+                }
+                _ => {
+                    continue;
+                }
+            };
+            // check for both bool (python class) and 'bool' (string annotation)
+            if ["bool", "'bool'"].iter().any(|v| v == &hint) {
```
Comment (Stranger6667) [inline src/flake8_boolean_trap/plugins.rs:44]: Would `.contains(&hint)` work here instead?

Comment (Stranger6667) [reply src/flake8_boolean_trap/plugins.rs:44]: Though, this would be faster

```rust
&hint == "bool" || &hint == "'bool'"
```

Comment (Stranger6667) [reply src/flake8_boolean_trap/plugins.rs:44]: See the comment above - maybe this approach is not needed

File: src/flake8_boolean_trap/plugins.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    for arg in arguments.args.iter().chain(arguments.posonlyargs.iter()) {
+        let hint;
+        if arg.node.annotation.is_none() {
+            continue;
+        }
+
+        if let Some(expr) = &arg.node.annotation {
+            match &expr.node {
+                ExprKind::Name { id, .. } => {
+                    hint = id.to_string();
```
Comment (Stranger6667) [inline src/flake8_boolean_trap/plugins.rs:34]: maybe `to_string` could be avoided altogether. E.g. `hint` could be a boolean that stores whether the annotation is `bool` / `"bool"`.

Here we can compare `&id` to `"bool"` and in the "Constant" branch we can match on `Constant::Str` similarly.

File: src/flake8_boolean_trap/plugins.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
+use rustpython_ast::{Arguments, ExprKind};
+use rustpython_parser::ast::{Constant, Expr, Located};
+
+use crate::ast::types::Range;
+use crate::check_ast::Checker;
+use crate::checks::{Check, CheckKind};
+
+fn is_boolean_arg(arg: &Located<ExprKind>) -> bool {
```
Comment (charliermarsh) [inline src/flake8_boolean_trap/plugins.rs:8]: Nit: Should be able to use `arg: &Expr` here (and elsewhere in this file).

File: src/flake8_boolean_trap/plugins.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            }
+        }
+    }
+}
+
+pub fn check_boolean_default_value_in_function_definition(
+    checker: &mut Checker,
+    arguments: &Arguments,
+) {
+    for arg in arguments.defaults.iter() {
```
Comment (charliermarsh) [inline src/flake8_boolean_trap/plugins.rs:58]: Do you need to check `kw_defaults` here too?

File: src/flake8_boolean_trap/plugins.rs (hunk 2 above)
Comment (charliermarsh) [reply src/flake8_boolean_trap/plugins.rs:34]: Yeah that'd be preferable, I think.

Like:
```rust
let hint = match &expr.node {
  ExprKind::Name { id, .. } => id == "bool",
  ExprKind::Constant { value, .. } => ..
  _ => false,
};
```

**Review by charliermarsh — COMMENTED** [review]
Generally looks good! A couple small comments.

File: src/flake8_boolean_trap/plugins.rs (hunk 4 above)
Comment (pwoolvett) [reply src/flake8_boolean_trap/plugins.rs:58]: kw_defaults are fine as booleans (in fact that would be a valid ast fix in an upcoming release), as they do not allow positional calls. MWE:

```py
def fun(a, *, b=True):
  ...
fun(4, True) # raises TypeError
```

File: src/flake8_boolean_trap/plugins.rs (hunk 2 above)
Comment (pwoolvett) [reply src/flake8_boolean_trap/plugins.rs]: corrected, please advise (tests passing, no lint suggestions)

File: src/flake8_boolean_trap/plugins.rs (hunk 1 above)
Comment (pwoolvett) [reply src/flake8_boolean_trap/plugins.rs]: adressed with charlie's suggestion

File: src/flake8_boolean_trap/plugins.rs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+fn add_if_boolean(checker: &mut Checker, arg: &Expr, kind: CheckKind) {
+    if is_boolean_arg(arg) {
+        checker.add_check(Check::new(kind, Range::from_located(arg)));
+    }
+}
+
+pub fn check_positional_boolean_in_def(checker: &mut Checker, arguments: &Arguments) {
+    for arg in arguments.posonlyargs.iter().chain(arguments.args.iter()) {
+        // let hint;
```
Comment (Stranger6667) [inline src/flake8_boolean_trap/plugins.rs:26]: This line seems outdated

Comment (pwoolvett) [reply src/flake8_boolean_trap/plugins.rs:26]: you mean just the comment, right?

Comment (Stranger6667) [reply src/flake8_boolean_trap/plugins.rs:26]: Yes, I guess it could be removed :)

Comment (pwoolvett) [reply src/flake8_boolean_trap/plugins.rs]: done

**Comment by charliermarsh** [comment]
Looks great! I'll merge this later today.

**Pull request merged**